File: packages/js-core/src/client-internal/pairing-session-manager.ts
import type { PolycentricClient } from '../polycentric-client';
import * as Proto from '../proto/v2';
import { bytesToHex } from '../utils/hex';

export interface ActivePairingSession {
  code: string;
  identityKey: string;
  createdAt: Date;
  expiresAt: Date;
  signedBy: Proto.PublicKey;
  claimers: Proto.PublicKey[];
  server: string;
}

export interface PairingSessionView {
  session: Proto.PairingSession;
  claimerPubkeys: Proto.PublicKey[];
  pairingSession: {
    issuerIdentity: string;
    createdAt: Date;
    expiresAt: Date;
    signedBy: Proto.PublicKey;
  };
  claimers: Proto.PublicKey[];
}

export class PairingSessionManager {
  constructor(private readonly client: PolycentricClient) {}

  private async signMessage(
    messageBytes: Uint8Array,
  ): Promise<Proto.SignedMessage> {
    if (!this.client.currentKeyPair) throw new Error('No active key pair');
    const signature = await this.client.crypto.sign(
      this.client.currentKeyPair.privateKey.key,
      messageBytes,
      this.client.currentKeyPair.keyType,
    );
    return Proto.SignedMessage.create({
      signature,
      messageBytes,
      publicKey: this.client.currentKeyPair.publicKey,
    });
  }

  /**
   * Creates a signed pairing session and registers it on the target server.
   * @param identityKey Identity key to embed in the session payload.
   * @param server Server URL that will store and serve the session.
   * @returns Session metadata returned from the server.
   */
  async createPairingSessionOnServer(
    identityKey: string,
    server: string,
  ): Promise<ActivePairingSession> {
    const pairingSessionBytes = Proto.InitialPairingSession.toBinary(
      Proto.InitialPairingSession.create({
        issuerIdentity: identityKey,
        timestamp: BigInt(Date.now()),
      }),
    );
    const signedMessage = await this.signMessage(pairingSessionBytes);

    const sessionBytes = await this.client.core.createPairingSession(
      server,
      Proto.SignedMessage.toBinary(signedMessage).buffer as ArrayBuffer,
    );
    const session = Proto.PairingSession.fromBinary(
      new Uint8Array(sessionBytes),
    );

    return {
      code: bytesToHex(signedMessage.signature),
      identityKey: session.issuerIdentity,
      createdAt: new Date(Number(session.createdAt)),
      expiresAt: new Date(Number(session.expiresAt)),
      signedBy: session.signedBy!,
      claimers: [],
      server,
    };
  }

  async getPairingSessionStatus(
    pairingSessionSignature: string,
    server?: string,
  ): Promise<PairingSessionView> {
    const targetServer = server ?? this.client.servers[0];
    if (!targetServer) throw new Error('No servers configured');

    const sessionBytes = await this.client.core.getPairingSession(
      targetServer,
      pairingSessionSignature,
    );
    const session = Proto.PairingSession.fromBinary(
      new Uint8Array(sessionBytes),
    );
    return {
      session,
      claimerPubkeys: [...session.claimerPubkeys],
      pairingSession: {
        issuerIdentity: session.issuerIdentity,
        createdAt: new Date(Number(session.createdAt)),
        expiresAt: new Date(Number(session.expiresAt)),
        signedBy: session.signedBy!,
      },
      claimers: [...session.claimerPubkeys],
    };
  }

  async joinPairingSession(
    pairingSessionSignature: string,
    server: string,
  ): Promise<PairingSessionView> {
    const bodyBytes = Proto.JoinPairingSessionBody.toBinary(
      Proto.JoinPairingSessionBody.create({
        pairingSessionSignature,
      }),
    );
    const signedMessage = await this.signMessage(bodyBytes);

    const sessionBytes = await this.client.core.joinPairingSession(
      server,
      Proto.SignedMessage.toBinary(signedMessage).buffer as ArrayBuffer,
    );
    const session = Proto.PairingSession.fromBinary(
      new Uint8Array(sessionBytes),
    );
    return {
      session,
      claimerPubkeys: [...session.claimerPubkeys],
      pairingSession: {
        issuerIdentity: session.issuerIdentity,
        createdAt: new Date(Number(session.createdAt)),
        expiresAt: new Date(Number(session.expiresAt)),
        signedBy: session.signedBy!,
      },
      claimers: [...session.claimerPubkeys],
    };
  }
}

File: packages/kotlin-core/core/src/main/kotlin/org/futo/polycentric/core/PairingSessionManager.kt
package org.futo.polycentric.core

import okio.ByteString.Companion.toByteString
import polycentric.v2.InitialPairingSession
import polycentric.v2.JoinPairingSessionBody
import polycentric.v2.PairingSession
import polycentric.v2.PublicKey
import polycentric.v2.SignedMessage

/**
 * Port of js-core `client-internal/pairing-session-manager.ts` — the
 * device-pairing handshake. Transport is entirely in the core
 * (`createPairingSession` / `getPairingSession` / `joinPairingSession`);
 * this manager builds and signs the payloads.
 *
 * Flow: an existing device creates a session for its identity; the new
 * device joins it with its own key; the existing device polls the
 * session, sees the claimer key, and calls
 * `IdentityManager.addSigningKey` — the new device then `claim()`s.
 */
class PairingSessionManager(private val client: PolycentricClient) {

    class ActivePairingSession(
        /** Hex signature of the signed InitialPairingSession — the session id/code. */
        val code: String,
        val identityKey: String,
        val createdAt: Long,
        val expiresAt: Long,
        val signedBy: PublicKey,
        val claimers: List<PublicKey>,
        val server: String,
    )

    class PairingSessionView(
        val session: PairingSession,
        val issuerIdentity: String,
        val createdAt: Long,
        val expiresAt: Long,
        val signedBy: PublicKey?,
        /** Public keys that have joined the session so far. */
        val claimers: List<PublicKey>,
    )

    private suspend fun signMessage(messageBytes: ByteArray): SignedMessage {
        val keyPair = client.currentKeyPair ?: throw NoActiveKeyPairException()
        val signature = client.crypto.sign(keyPair.privateKey, messageBytes, keyPair.keyType)
        return SignedMessage(
            signature = signature.toByteString(),
            message_bytes = messageBytes.toByteString(),
            public_key = keyPair.toPublicKeyProto(),
        )
    }

    /**
     * Creates a signed pairing session and registers it on the target
     * server. The returned `code` (signature hex) is what the joining
     * device presents.
     */
    suspend fun createPairingSessionOnServer(
        identityKey: String,
        server: String,
    ): ActivePairingSession {
        val payloadBytes = InitialPairingSession.ADAPTER.encode(
            InitialPairingSession(
                issuer_identity = identityKey,
                timestamp = System.currentTimeMillis(),
            ),
        )
        val signedMessage = signMessage(payloadBytes)

        val sessionBytes = client.core.createPairingSession(
            server,
            SignedMessage.ADAPTER.encode(signedMessage),
        )
        val session = PairingSession.ADAPTER.decode(sessionBytes)

        return ActivePairingSession(
            code = signedMessage.signature.hex(),
            identityKey = session.issuer_identity,
            createdAt = session.created_at,
            expiresAt = session.expires_at,
            signedBy = session.signed_by ?: throw PolycentricException("Session missing signed_by"),
            claimers = emptyList(),
            server = server,
        )
    }

    /** Fetch a session's current state (poll while waiting for claimers). */
    suspend fun getPairingSessionStatus(
        pairingSessionSignature: String,
        server: String? = null,
    ): PairingSessionView {
        val targetServer = server ?: client.servers.firstOrNull()
            ?: throw PolycentricException("No servers configured")

        val sessionBytes = client.core.getPairingSession(targetServer, pairingSessionSignature)
        return PairingSession.ADAPTER.decode(sessionBytes).toView()
    }

    /** Join a session as a claimer, signing the join body with our key. */
    suspend fun joinPairingSession(
        pairingSessionSignature: String,
        server: String,
    ): PairingSessionView {
        val bodyBytes = JoinPairingSessionBody.ADAPTER.encode(
            JoinPairingSessionBody(pairing_session_signature = pairingSessionSignature),
        )
        val signedMessage = signMessage(bodyBytes)

        val sessionBytes = client.core.joinPairingSession(
            server,
            SignedMessage.ADAPTER.encode(signedMessage),
        )
        return PairingSession.ADAPTER.decode(sessionBytes).toView()
    }

    private fun PairingSession.toView() = PairingSessionView(
        session = this,
        issuerIdentity = issuer_identity,
        createdAt = created_at,
        expiresAt = expires_at,
        signedBy = signed_by,
        claimers = claimer_pubkeys,
    )
}
